# EX1 Descriptive statistics
The mean is one of typical descriptive statistics. The mean is used for describing the central tendency of data.

When the data consists of $n$ numerical elements, the mean can be calculated as follows. 

$\frac{1}{n}\Sigma^{n}_{i=1}x_{i}$,

where $x_{i}$ represents each element of the data.

In the following,  implement Python code to calculate the mean, maximum, and minimum values given numerical data as a list.

Points
- Q1 5P
- Q2 5P

## Q1
Complete the `avg_max_min` function, which takes as input a list of any length with non-negative integers (at most 10,000,000) as elements and returns a list with the following as elements.
- The mean value of the elements of the input list
- The maximum value among the elements of the input list
- The minimum value among the elements of the input list

In [3]:
def avg_max_min(input_list):
  avg_value=0 # initial mean value
  max_value=-1 # initial max value
  min_value= 10**7 # initial min value
  sum_value=0 # total value of the elements
  results=[] # list to be returned
    
  for val in input_list:
    sum_value = sum_value + val
        
    if val>max_value:
      max_value = val
            
    if val<min_value:
      min_value = val
        
  avg_value =  sum_value/len(input_list)
    
    
  results.append(avg_value) 
  results.append(max_value)
  results.append(min_value)
  return results  

Once the `avg_max_min` function is complete, run the following cell to see how it works.

In [4]:
avg_max_min([30,39,29,95,70,67,29,56,45,68])

[52.8, 95, 29]

Make sure that your function returns `[52.8, 95, 29]`. 

## Q2
Complete the `population_stat` function:
- first, read a "SSDSE.csv" file.
- create a list of mean, maximum and minimum values for each column,  "total population", "total population (male)", "total population (female)".
- The lists becomes the values of the dictionary to be returned, where its keys are "total", "male", and "female".

you may use the `avg_max_min` function above.


```Python
## SSDSE.csv file
code, prefecture, municipality, A1101,	A110101, A110102, ...
地域コード, 都道府県, 市区町村, 人口総数, 人口総数（男）, 人口総数（女）, ...
R01100, 北海道, 札幌市, 1952356, 910614, 1041742
R01202, 北海道, 函館市, 265979, 120376, 145603
R01203, 北海道, 小樽市, 121924, 54985, 66939
...

## A dictionary to be returned
{'total': ['人口総数の平均', '人口総数の最大値', '人口総数の最小値'],
 'male': ['人口総数（男）の平均', '人口総数（男）の最大値', '人口総数（男）の最小値'],
 'female': ['人口総数（女）の平均', '人口総数（女）の最大値', '人口総数（女）の最小値']}
```

### Reference

- [教育用標準データセット](https://www.nstac.go.jp/SSDSE/)

In [5]:
# Run this cell to download the data
!wget https://raw.githubusercontent.com/UTDataMining/2022A/master/ex1/SSDSE.csv

--2022-10-13 10:53:53--  https://raw.githubusercontent.com/UTDataMining/2022A/master/ex1/SSDSE.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 867732 (847K) [text/plain]
Saving to: ‘SSDSE.csv’

SSDSE.csv           100%[===================>] 847.39K  --.-KB/s    in 0.008s  

2022-10-13 10:53:53 (109 MB/s) - ‘SSDSE.csv’ saved [867732/867732]



In [9]:
import csv

def population_stat():
  total=[] # initial list of total population
  male=[] # initial list of male population
  female=[] # initial list of female population
    
  with open('SSDSE.csv', 'r', encoding="utf-8") as handler: 
    csv_reader = csv.reader(handler) 
    for row in csv_reader: 
      if csv_reader.line_num > 2: 
          total.append(int(row[3]))
          male.append(int(row[4]))
          female.append(int(row[5]))
    
    total_stat = avg_max_min(total)
    male_stat = avg_max_min(male)
    female_stat = avg_max_min(female)
    
    results={"total":total_stat, "male":male_stat, "female":female_stat}
    return results

Once the `poplation_stat` function is complete, run the following cell to see how it works.

In [10]:
population_stat()

{'total': [73001.00229753015, 3724844, 0],
 'male': [35520.814474439976, 1855985, 0],
 'female': [37480.18782309018, 1868859, 0]}

Make sure your function returns`{'total': [73001.00229753015, 3724844, 0],
 'male': [35520.814474439976, 1855985, 0],
 'female': [37480.18782309018, 1868859, 0]}`.

The order of the keys can be different.
 

# Code Testing

In [11]:
# Run this cell
!pip install prog_edu_assistant_tools
import re
import sys
import jinja2
from IPython.core import display
from google.colab import _message as google_message
from prog_edu_assistant_tools.magics import report, autotest, CaptureOutput
from prog_edu_assistant_tools.check import Check

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.6 MB 29.2 MB/s 


## Q1

In [12]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q1')

## Q2

In [13]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2')